### DISCUSSION TITLE
Glyphs with a custom data set

### DISCUSSION THREAD

**Discussion by kimjohn1** [SCIInstitute/SCIRun#2265] [Q&A] [answered]
SCIRun experts

I'm wrapping up a project that visualizes position data generated by a tiny Inertial Measurement Unit (IMU).  I (separately) parse the raw data from the device to produce a text file that has 3 values per line of text.  The 3 values are X, Y, Z position, and the lines are recorded at 1 second intervals.  I am able to input that text file and render the position data as a series of spheres that represent the path taken and recorded by the IMU.  So far so good.

I have access to the X, Y, Z velocity values that were recorded at the same time as the position, so I would like to include the velocity at each position point in the visualization, and show the velocity vector as a glyph.  So far I have had zero success.  Can you tell me how I should format a data file, or files, to include position and velocity, input them to SCIRun and process them to render the velocity vector at each position point as a glyph?

Kim

**Comment by jessdtate** [comment, marked as answer]
There are a few ways to do this.  The simplest way may be to save the vectors as a text file, then load that with Readmatrix and use setfielddata to set the vectors onto the pointcloud.  ShowFieldGlyphs should work after that.  

Another way to save the data to load in the vector field and positions at the same time is to save it in matlab format as a structure.  The locations should be set in a field called 'node' or 'pts'.  The vectors should be saved in a field called 'field'.  

one way to make sure that you have vector data assigned to the field is to use reportfieldinfo, or to click on a pipe and press the 'i' key, or to cmd(crt) click on an output port.  Look for a vector<vector> for vector data, vs vector<double> for scalar data.

**Reply by jessdtate** [reply]
<img width="420" alt="data_info_double" src="https://user-images.githubusercontent.com/3484631/115479309-64866e80-a205-11eb-8029-cc629f3853f5.png">
<img width="417" alt="data_info" src="https://user-images.githubusercontent.com/3484631/115479311-65b79b80-a205-11eb-9dc5-6fd38b13bfc3.png">

here are some images highlighting the difference.

**Reply by kimjohn1** [reply]
@jessdtate 
Excellent suggestions.  Thanks!
    I could only get ```vector<double>``` out of SetFieldData, but that got me thinking.  I tried using CalculateFieldData, supplying it with XYZ position and XYZ velocity from separate files
![dots_net](https://user-images.githubusercontent.com/61957753/115528073-eea6f500-a24e-11eb-9dcb-74db4cfbb0bf.png)
With a simple ```RESULT=Vector(X, Y, Z);``` calculation, I get a vector output that shows glyphs.  Huzzah.  I don't know what calculation to use in CalculateFieldData, to get an output that takes the XYZ position data input for positioning, and the XYZ velocity input for the vector output.  Any ideas for that last bit?

Kim

**Reply by jessdtate** [reply]
I'm not sure what you are asking.  `RESULT=vector(X,Y,Z);` should give you what I think you are asking, ie, a set of points with vectors that where the position are the same as the vector.  Another way to achieve this with CalculateFieldData is `RESULT = POS;`.

**Reply by jessdtate** [reply]
I think I understand now.  You want some kind of operation that is dependent on one of the position dimensions and the velocity vector?  Something like this would work:
```
velocity = vector(X,Y,Z);
RESULT = X*velocity;
```
This will yield the position vector (velocity) scaled by the X direction.

**Reply by jessdtate** [reply]
You can look at the parser help (a button near the bottom of the module UI) for more operations that you might use.

**Reply by kimjohn1** [reply]
@jessdtate 
These are good suggestions and I have tried them, but not getting what I want yet.  I believe you are correct to focus on the calculation being used in CalculateFieldData.  I think I am not using a correct calculation there to produce a properly constructed field for the output.  I've read through the Parser description a couple of times.  It looks like the CalculateFieldData module is expecting a field as input, but the 2 inputs I provide are the position points provided by AutoReadFile1 and velocity vectors for each of the points provided by AutoReadFile2.  Both of these inputs show as ```vector <double>``` for the data type.  The output from CalculateFieldData shows as ```vector <vector>``` for the field data type, but that output vector only contains data from whatever I have connected to the first input port.  I need to figure out the function that would take data from the two inputs and create a vector field RESULT that includes both position data (from the one input) and velocity data (from the other input).  I'll keep working with it.
Kim

**Reply by kimjohn1** [reply]
@jessdtate 
Got it.  The key was in your first reply, first paragraph, where you suggest using SetFieldData.  I was using SetFieldData to format the input position data (a scalar field), and was using CalculateFieldData to try to construct a vector field from the separately input velocity vector data.  When I went back and read through the description of the SetFieldData module, I realized I could feed the input velocity vector data as well as the position data to SetFieldData.  
![glyph_net](https://user-images.githubusercontent.com/61957753/115812411-91c25080-a3ae-11eb-89b8-07b439dc9c2f.png)
Works like a champ.

I verified that the glyphs were correctly positioned and oriented by manually constructing a data set that contained a random distribution of points on the surface of a sphere, and input that data set as the position data and as the velocity vector data.  Correctly placed and oriented glyphs should emanate from each position point and be pointing directly away from the center of the sphere:
![sphere](https://user-images.githubusercontent.com/61957753/115812831-4bb9bc80-a3af-11eb-9df5-6c8d211aa58b.png)
I'll mark this one as answered.  Thanks.

Kim

**Comment by kimjohn1** [comment]
The issue has been resolved.  See Reply above.
Kim
